packages/aws-sdk-client-mock/src/awsClientSpy.ts

```diff
@@ -0,0 +1,118 @@
+import {Client, MetadataBearer} from '@smithy/types';
+import {match, SinonSpyCall, SinonSpy} from 'sinon';
+import {AwsCommand} from './commonTypes';
+import {spyClient} from './spyClient';
+
+/**
+ * Type for {@link AwsSpy} class,
+ * but with the AWS Client class type as the only generic parameter.
+ *
+ * @example
+ * ```ts
+ * const sns = new SNSCLient({});
+ * let snsSpy: AwsClientSpy<SNSClient>;
+ * snsSpy = spyClient(sns);
+ * ```
+ */
+export type AwsClientSpy<TClient> =
+    TClient extends Client<infer TInput, infer TOutput, infer TConfiguration> ? AwsSpy<TInput, TOutput, TConfiguration> : never;
+
+/**
+ * Wrapper on the mocked `Client#send()` method, allowing you to inspect
+ * calls to it.
+ *
+ * Without any configuration, `Client#send()` invocation returns `undefined`.
+ *
+ * To define resulting variable type easily, use {@link AwsClientSpy}.
+ */
+export class AwsSpy<TInput extends object, TOutput extends MetadataBearer, TConfiguration> {
```

Comment:
Because Sinon Stub extends the Spy, we could do the same here - AwsStub extend AwsSpy. The `call()`, `calls()`, and `commandCalls()` could be defined only in AwsSpy.

Reply:
This makes sense in theory, but in practice, the type shenanigans needed to type the `send` value and the `reset` call so they return different types makes this extremely difficult (and probably very confusing to downstream users). 

I did experiment with this, but I think it's better to have separate definitions and avoid the complexity of inheritance.

Reply:
That said, I'll take another look as it may make sense from a testing perspective to have these inherited

Reply:
Okay, thanks, let me know if you don't find a sensible way to implement this, I will take a look then or we will leave it as it is 👍